modules/tensor_mechanics/src/materials/CZM/CZM_3DCMaterial.C

```diff
@@ -0,0 +1,166 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "Assembly.h"
+#include "CZM_3DCMaterial.h"
+#include "RotationMatrix.h"
+
+registerMooseObject("TensorMechanicsApp", CZM_3DCMaterial);
+
+template <>
+InputParameters
+validParams<CZM_3DCMaterial>()
+{
+  InputParameters params = validParams<CZMMaterialBase>();
+  params.addClassDescription("3DC cohseive law model, no damage");
+  params.addRequiredParam<std::vector<Real>>(
+      "DeltaU0",
+      "a vector containing the displacement value at which maximum"
+      " traction occurs for the normal(1st) and tangential(2nd) "
+      " direction.");
+  params.addRequiredParam<std::vector<Real>>("MaxAllowableTraction",
+                                             "a vector containing the maximum allowed traction"
+                                             "for the normal(1st) and tangential(2nd) direction.");
+  params.addClassDescription("Simple Exponential cohseive law model, with damage");
+  return params;
+}
+
+CZM_3DCMaterial::CZM_3DCMaterial(const InputParameters & parameters)
+  : CZMMaterialBase(parameters),
+    _deltaU0(getParam<std::vector<Real>>("DeltaU0")),
+    _maxAllowableTraction(getParam<std::vector<Real>>("MaxAllowableTraction"))
+
+{
+  // check inputs
+  if (_deltaU0.size() != 2)
+    mooseError("CZMLaw3DC: the parameter DeltaU0 requires 2 components, " +
+               std::to_string(_deltaU0.size()) + " provided.");
+  if (_maxAllowableTraction.size() != 2)
+    mooseError("CZMLaw3DC: the parameter MaxAllowableTraction"
+               "requires 2 components," +
+               std::to_string(_maxAllowableTraction.size()) + " provided.");
+
+  // copying component 2 of _deltaU0 and _maxAllowableTraction to ease
+  // calculations
+  const_cast<std::vector<Real> &>(_deltaU0).push_back(_deltaU0[1]);
+  const_cast<std::vector<Real> &>(_maxAllowableTraction).push_back(_maxAllowableTraction[1]);
+}
+
+RealVectorValue
+CZM_3DCMaterial::computeLocalTraction() const
+{
+  RealVectorValue traction_local;
+
+  // convention N, T, S
+  Real temp, X, expX, A_i, B_i;
```

Comment:
can you use a different name for temp? we have been use temp for temperature in many places.